Cython/Utility/TypeConversion.c

```diff
@@ -286,20 +291,61 @@ static CYTHON_INLINE const char* __Pyx_PyObject_AsStringAndSize(PyObject* o, Py_
         return PyByteArray_AS_STRING(o);
 #else
         *length = PyByteArray_Size(o);
-        if (*length == -1) return NULL;
+        if (*length == -1) goto error;
         return PyByteArray_AsString(o);
 #endif
 
-    } else
-    {
+    } else if (PyBytes_Check(o)) {
         char* result;
         int r = PyBytes_AsStringAndSize(o, &result, length);
         if (unlikely(r < 0)) {
-            return NULL;
+            goto error;
         } else {
             return result;
         }
+    } else {
+        char* result;
+        int r;
+#if !CYTHON_COMPILING_IN_LIMITED_API || __PYX_LIMITED_VERSION_HEX >= 0x030B0000
+        // Check that `o`:
+        // * Supports the Python Buffer Protocol
+        // * There is no need to keep the buffer around (as we cannot return it)
+        PyBufferProcs *pb = Py_TYPE(o)->tp_as_buffer;
+        if (pb == NULL || pb->bf_getbuffer == NULL || pb->bf_releasebuffer != NULL) {
+            goto error;
+        }
+
+        // Try to acquire buffer from `o`
+        Py_buffer view;
+        r = PyObject_GetBuffer(o, &view, PyBUF_SIMPLE);
+        if (unlikely(r < 0)) {
+            goto error;
+        } else {
+            result = (char*) view.buf;
+            *length = view.len;
+            PyBuffer_Release(&view);  // Release to clean up buffer (decref object `o`)
```

Comment:
We still release the buffer as it does some cleanup of the buffer object itself (namely [decref-ing the associated object]( https://github.com/python/cpython/blob/52b5eb95b770fa00ebbd449ba40cab4a0e7c7df7/Objects/abstract.c#L817-L818 ))

Thought it better to call the standard function here instead of bypassing it by doing our own version of this cleanup